rpc/debug/trace.go

```diff
@@ -0,0 +1,196 @@
+		return nil, err
+	}
+
+	txResult, err := blockstore.GetTxResult(block.Block.Data.Txs[txIndex].Hash())
+	if err != nil {
+		return nil, err
+	}
+
+	if err := app.SetTracer(tracer, false); err != nil {
+		return nil, err
+	}
+
+	result := app.DeliverTx(block.Block.Data.Txs[txIndex])
+	match, err := resultsMatch(txResult.TxResult, result)
+	if !match {
+		return nil, err
+	}
+
+	switch tracer := tracer.(type) {
+	case *vm.StructLogger:
+		return &ethapi.ExecutionResult{
+			Gas:         0,
+			Failed:      result.Code != abci.CodeTypeOK,
+			ReturnValue: fmt.Sprintf("%x", result),
+			StructLogs:  ethapi.FormatLogs(tracer.StructLogs()),
+		}, nil
+	case *tracers.Tracer:
+		return tracer.GetResult()
+	default:
+		return nil, errors.New(fmt.Sprintf("bad tracer type %T", tracer))
+	}
+}
+
+func StorageRangeAt(
+	app loomchain.Application,
+	blockstore store.BlockStore,
+	address, begin []byte,
+	startBlockNumber, targetBlockNumber, txIndex int64,
+	maxResults int,
+) (results JsonStorageRangeResult, err error) {
+	defer func() {
+		if r := recover(); r != nil {
+			err = errors.Errorf("loomchain panicked %v", r)
+		}
+	}()
+
+	block, err := runTxsTo(&app, blockstore, startBlockNumber, targetBlockNumber, txIndex, true)
+	if err != nil {
+		return JsonStorageRangeResult{}, err
+	}
+
+	storeState := loomchain.NewStoreState(
+		context.Background(),
+		app.Store,
+		ttypes.TM2PB.Header(&block.Block.Header),
+		block.BlockMeta.BlockID.Hash,
+		app.GetValidatorSet,
+	)
+	ethDb := evm.NewLoomEthdb(storeState, nil)
+	root := app.Store.Get(util.PrefixKey(evm.VmPrefix, evm.RootKey))
+
+	stateDb, err := state.New(common.BytesToHash(root), state.NewDatabase(ethDb))
+	if err != nil {
+		return JsonStorageRangeResult{}, err
+	}
+	st := stateDb.StorageTrie(common.BytesToAddress(address))
+	result, err := eth.StorageRangeAt(st, begin, maxResults)
+	if err != nil {
+		return JsonStorageRangeResult{}, err
+	}
+
+	return JsonStorageRangeResult{
+		StorageRangeResult: result,
+		Complete:           result.NextKey == nil,
+	}, nil
+}
+
+func runTxsTo(app *loomchain.Application, blockstore store.BlockStore, startHeight, height, index int64, includeEnd bool) (*ctypes.ResultBlock, error) {
+	for h := startHeight; h <= height; h++ {
+		resultBlock, err := blockstore.GetBlockByHeight(&h)
+		if err != nil {
+			return nil, errors.Wrapf(err, "getting block information at height %v", h)
+		}
+
+		_ = app.BeginBlock(requestBeginBlock(*resultBlock))
+
+		if h != height {
+			for i := 0; i < len(resultBlock.Block.Data.Txs); i++ {
+				_ = app.DeliverTx(resultBlock.Block.Data.Txs[i])
+			}
+		} else {
+			for i := 0; i < len(resultBlock.Block.Data.Txs); i++ {
+				if i != int(index) {
+					_ = app.DeliverTx(resultBlock.Block.Data.Txs[i])
+				} else {
+					if includeEnd {
+						_ = app.DeliverTx(resultBlock.Block.Data.Txs[i])
+					}
+					return resultBlock, nil
+				}
+			}
+		}
+
+		_ = app.EndBlock(requestEndBlock(h))
+		_ = app.Commit()
```

Comment:
Can you add a TODO comment here to say that we should really be matching the app hash returned here with the one that was stored in the next block, otherwise we can't really know if the replay was successful or not.